tests/unit/test_id_match.py

```diff
@@ -0,0 +1,297 @@
+        ),
+        (
+            ['1/*'],
+            ['1/a:x', '1/b:x', '1/c:x'],
+            []
+        ),
+        (
+            ['2/*'],
+            [],
+            ['2/*']
+        ),
+        (
+            ['*/*'],
+            ['1/a:x', '1/b:x', '1/c:x'],
+            []
+        ),
+        (
+            ['*/a'],
+            ['1/a:x'],
+            []
+        ),
+        (
+            ['*/z'],
+            [],
+            ['*/z']
+        ),
+        (
+            ['*/*:x'],
+            ['1/a:x', '1/b:x', '1/c:x'],
+            [],
+        ),
+        (
+            ['*/*:y'],
+            [],
+            ['*/*:y'],
+        ),
+    ]
+)
+def test_filter_ids_task_mode(task_pool, ids, matched, not_matched):
+    """Ensure tasks are returned in task mode."""
+    pool = task_pool(
+        {
+            1: ['1/a:x', '1/b:x', '1/c:x']
+        },
+        {}
+    )
+
+    _matched, _not_matched = filter_ids([pool], ids)
+    assert [itask.id_ for itask in _matched] == matched
+    assert _not_matched == not_matched
+
+
+@pytest.mark.parametrize(
+    'ids,matched,not_matched',
+    [
+        (
+            ['1/a'],
+            [1],
+            [],
+        ),
+        (
+            ['1/*'],
+            [1],
+            [],
+        ),
+        (
+            ['1/*:x'],
+            [1],
+            [],
+        ),
+        (
+            ['1/*:y'],
+            [],
+            ['1/*:y'],
+        ),
+        (
+            ['*/*:x'],
+            [1],
+            [],
+        ),
+        (
+            ['1/z'],
+            [],
+            ['1/z'],
+        ),
+        (
+            ['1'],
+            [1],
+            [],
+        ),
+        (
+            ['3'],
+            [],
+            ['3'],
+        ),
+    ]
+)
+def test_filter_ids_cycle_mode(task_pool, ids, matched, not_matched):
+    """Ensure cycle poinds are returned in cycle mode."""
+    pool = task_pool(
+        {
+            1: ['1/a:x', '1/b:x'],
+            2: ['1/a:x'],
+            3: [],
+        },
+        {}
+    )
+
+    _matched, _not_matched = filter_ids([pool], ids, out=IDTokens.Cycle)
+    assert _matched == matched
+    assert _not_matched == not_matched
+
+
+def test_filter_ids_invalid(caplog):
+    """Ensure invalid IDs are handled elegantly."""
+    matched, not_matched = filter_ids([{}], ['#'])
+    assert matched == []
+    assert not_matched == ['#']
+    assert caplog.record_tuples == [
+        ('cylc', 30, 'No active tasks matching: #'),
+    ]
+    caplog.clear()
+    matched, not_matched = filter_ids([{}], ['#'], warn=False)
+    assert caplog.record_tuples == []
+
+
+def test_filter_ids_pattern_match_off(task_pool):
+    """Ensure filtering works when pattern matching is turned off."""
+    pool = task_pool(
+        {
+            1: ['1/a:x'],
+        },
+        {}
+    )
+
+    _matched, _not_matched = filter_ids(
+        [pool],
+        ['1/a'],
+        out=IDTokens.Task,
+        pattern_match=True,
```

Comment:
Seems wrong given the name + description of the test?
```suggestion
        pattern_match=False,
```

Reply:
Fixed.